From 203d1b5d392391df073af73b4a81e4f96afea12c Mon Sep 17 00:00:00 2001
From: oliverwen <oliverwen@google.com>
Date: Wed, 13 Jul 2016 17:10:05 -0700
Subject: [PATCH] CHROMIUM: smart antenna algorithm: bugfix: rx_ant reversion

 - fix the bug of rx ant selection not reverted to default
   antenna value

TEST=run throughput test with multiple clients, when tx_ant of clients
diverge, the rx_ant should be reverted back to default value.

BUG=chrome-os-partner:51964
CQ-DEPEND=CL:341557

Signed-off-by: oliverwen <oliverwen@google.com>
(cherry picked from commit 8dfcbdbe79fad2a0b0f50781871e537c28ea82b0)
(source: wireless-3.18)

Change-Id: I6d1dc9a335c7204cb39f4814c86cd25075c1983c
Reviewed-on: https://chromium-review.googlesource.com/341558
Commit-Ready: Yixiang Li <yixiang@google.com>
Tested-by: Oliver Wen <oliverwen@chromium.org>
Reviewed-by: Yixiang Li <yixiang@google.com>
---
 .../net/wireless-4.2/ath/ath10k/smart_ant.c   | 26 +++++++++++--------
 .../net/wireless-4.2/ath/ath10k/smart_ant.h   |  1 +
 2 files changed, 16 insertions(+), 11 deletions(-)

diff --git a/drivers/net/wireless/ath/ath10k/smart_ant.c b/drivers/net/wireless/ath/ath10k/smart_ant.c
index 773738f56b3..eef908e1c8e 100644
--- a/drivers/net/wireless/ath/ath10k/smart_ant.c
+++ b/drivers/net/wireless/ath/ath10k/smart_ant.c
@@ -592,17 +592,21 @@ static u8 smart_ant_sel_rx_ant(struct ath10k *ar,
 			   sa_info->num_sta_conneted, nsta_sel_ant);
 	}
 
-	if ((tinfo->sel_ant != sa_info->default_ant) &&
-	    (sa_info->num_sta_conneted == nsta_sel_ant)) {
-		sa_info->default_ant = tinfo->sel_ant;
-		status = ATH10K_SMART_ANT_ACT_RX_CFG;
+	/* When all stations select the same tx_ant, rx_ant follows tx_ant.
+	* Otherwise, rx_ant is set to be default_ant.
+	*/
+	if (sa_info->num_sta_conneted == nsta_sel_ant)
+		sa_info->rx_ant = tinfo->sel_ant;
+	else
+		sa_info->rx_ant = sa_info->default_ant;
 
-		if (ar->smart_ant_info.debug_level >=
-		    ATH10K_SMART_ANT_DBG_LVL_TRAIN_STATES) {
-			ath10k_dbg(ar, ATH10K_DBG_SMART_ANT,
-				   "Rx antenna is going to be config to %d\n",
-				   tinfo->sel_ant);
-		}
+	status = ATH10K_SMART_ANT_ACT_RX_CFG;
+
+	if (ar->smart_ant_info.debug_level >=
+	    ATH10K_SMART_ANT_DBG_LVL_TRAIN_STATES) {
+		ath10k_dbg(ar, ATH10K_DBG_SMART_ANT,
+			   "Rx antenna is going to be config to %d\n",
+			   sa_info->rx_ant);
 	}
 
 	tinfo->prev_sel_ant = tinfo->sel_ant;
@@ -1300,7 +1304,7 @@ static int smart_ant_config_rx(struct ath10k *ar,
 	}
 
 	cfg->type = ATH10K_SMART_ANT_TYPE_RX_CFG;
-	cfg->rx_ant_cfg.rx_ant = info->default_ant;
+	cfg->rx_ant_cfg.rx_ant = info->rx_ant;
 
 	if (ar->smart_ant_info.debug_level >=
 	    ATH10K_SMART_ANT_DBG_LVL_TOP_DECISION) {
diff --git a/drivers/net/wireless/ath/ath10k/smart_ant.h b/drivers/net/wireless/ath/ath10k/smart_ant.h
index 06eea6ddb83..8c4b2049dde 100644
--- a/drivers/net/wireless/ath/ath10k/smart_ant.h
+++ b/drivers/net/wireless/ath/ath10k/smart_ant.h
@@ -472,6 +472,7 @@ struct ath10k_smart_ant_info {
 	u32 enabled_feedback;
 	u8 mode;
 	u8 default_ant;
+	u8 rx_ant;
 	u8 num_fallback_rate;
 	u8 num_sta_per_ant[ATH10K_SMART_ANT_COMB_MAX];
 	u16 num_sta_conneted;
-- 
2.43.0

